restarter/hot-restarter.py

```diff
@@ -5,18 +5,69 @@
 import sys
 import time
 
+# The number of seconds to wait for children to gracefully exit after
+# propagating SIGTERM before force killing children.
+# NOTE: If using a shutdown mechanism such as runit's `force-stop` which sends
+# a KILL after a specified timeout period, it's important to ensure that this
+# constant is smaller than the KILL timeout
+TERM_WAIT_SECONDS = 30
+
 restart_epoch = 0
 pid_list = []
 
-def force_kill_all_children():
-  """ Iterate through all known child processes and force kill them. In the future we might consider
-      possibly giving the child processes time to exit but this is fine for now. If someone force kills
-      us and does not clean the process tree this will leave child processes around unless they choose
-      to end themselves if their parent process dies. """
+def term_all_children():
+  """ Iterate through all known child processes, send a TERM signal to each of
+  them, and then wait up to TERM_WAIT_SECONDS for them to exit gracefully,
+  exiting early if all children go away. If one or more children have not
+  exited after TERM_WAIT_SECONDS, they will be forcibly killed """
 
   # First uninstall the SIGCHLD handler so that we don't get called again.
   signal.signal(signal.SIGCHLD, signal.SIG_DFL)
 
+  global pid_list
+  for pid in pid_list:
+    print "sending TERM to PID={}".format(pid)
+    try:
+        os.kill(pid, signal.SIGTERM)
+    except:
```

Comment:
Nit: prefer not catch-all exception handling.

Reply:
are there certain exceptions that you think should be retried here? or do you want the script to exit? For the containers use case it's fairly important that the python process doesn't exit until the children have exited or else we will leak container state in tmpfs